common/cache/budget.go

```diff
@@ -0,0 +1,1260 @@
+
+	usedBytes uint64 // atomic - total usage across all caches
+	usedCount int64  // atomic - total usage across all caches
+
+	// Per-cache usage tracking using sync.Map for lock-free operations
+	cacheUsage sync.Map // map[string]*Usage
+
+	// Optimistic tracking of active caches (usage > 0)
+	activeCacheCount int64 // atomic
+}
+
+// NewBudgetManager creates a new Manager.
+//
+// Parameters:
+//   - name: manager name for identification
+//   - maxBytes: bytes capacity semantics:
+//     < 0  => enforcement disabled (unlimited; still tracked)
+//     == 0 => caching fully disabled (deny all reserves)
+//     > 0  => enforcement enabled with that capacity
+//   - maxCount: count capacity semantics (same as maxBytes)
+//   - admission: AdmissionOptimistic (add-then-undo) or AdmissionStrict (CAS pre-check)
+//   - reclaimBackoff: optional sleep between reclaim attempts (defaults to ~100µs if zero)
+//   - scope: metrics scope for emitting metrics (can be nil)
+//   - logger: logger for diagnostic messages (can be nil)
+//   - softCapThreshold: optional percentage (0.0-1.0) defining how capacity is split between two tiers:
+//   - Free space tier: (threshold * capacity) - shared by all caches
+//   - Fair share tier: ((1 - threshold) * capacity) / activeCaches - allocated per cache
+//     The fair share capacity is equal to the fair share capacity ((1 - threshold) * capacity) divided by the number
+//     of active caches (caches with usage > 0).
+func NewBudgetManager(
+	name string,
+	maxBytes dynamicproperties.IntPropertyFn,
+	maxCount dynamicproperties.IntPropertyFn,
+	admission AdmissionMode,
+	reclaimBackoff time.Duration,
+	scope metrics.Scope,
+	logger log.Logger,
+	softCapThreshold dynamicproperties.FloatPropertyFn, // nil defaults to 1.0 (disabled)
+) Manager {
+	if scope == nil {
+		scope = metrics.NoopScope
+	}
+	if maxBytes == nil {
+		maxBytes = dynamicproperties.GetIntPropertyFn(-1) // unlimited if maxBytes is not provided
+	}
+	if maxCount == nil {
+		maxCount = dynamicproperties.GetIntPropertyFn(-1) // unlimited if maxCount is not provided
+	}
+	if softCapThreshold == nil {
+		softCapThreshold = dynamicproperties.GetFloatPropertyFn(1.0) // no threshold if softCapThreshold is not provided
+	}
+	if admission == 0 {
+		admission = AdmissionOptimistic
+	}
+
+	mgr := &manager{
+		name:                        name,
+		maxBytes:                    maxBytes,
+		maxCount:                    maxCount,
+		reclaimBackoff:              reclaimBackoff,
+		admission:                   admission,
+		scope:                       scope.Tagged(metrics.BudgetManagerNameTag(name)),
+		logger:                      logger,
+		enforcementSoftCapThreshold: softCapThreshold,
+	}
+
+	// Emit initial metrics
+	mgr.updateMetrics()
+
+	return mgr
+}
+
+// updateMetrics emits current state metrics
+func (m *manager) updateMetrics() {
+	// Emit capacity metrics
+	capacityBytes := m.CapacityBytes()
+	if capacityBytes != math.MaxUint64 {
```

Comment:
I don't think it's super likely you need the guard, as it'll wraparound if its exceeding maxint, though by that time whatever you're measuring is pretty big, so it's probably beyond any reasonable size that could be in memory for current computing. 

I know I was being annoying about it before, I think I was fixating on the wrong thing in practice. If you do want to put guards in place, they probably should be `<`